pandas/core/indexes/datetimes.py

```diff
@@ -252,19 +253,54 @@ def __new__(cls, data=None,
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
-        if not isinstance(data, (np.ndarray, Index, ABCSeries, DatetimeArray)):
-            # other iterable of some kind
-            if not isinstance(data, (list, tuple)):
+        if not hasattr(data, "dtype"):
+            # e.g. list, tuple
+            if np.ndim(data) == 0:
+                # i.e. generator
                 data = list(data)
-            data = np.asarray(data, dtype='O')
+            data = np.asarray(data)
+            copy = False
         elif isinstance(data, ABCSeries):
             data = data._values
 
-        # data must be Index or np.ndarray here
-        if not (is_datetime64_dtype(data) or is_datetimetz(data) or
-                is_integer_dtype(data) or lib.infer_dtype(data) == 'integer'):
-            data = tools.to_datetime(data, dayfirst=dayfirst,
-                                     yearfirst=yearfirst)
+        # By this point we are assured to have either a numpy array or Index
+
+        if is_float_dtype(data):
+            # Note: we must cast to datetime64[ns] here in order to treat these
+            #  as wall-times instead of UTC timestamps.
+            data = data.astype(_NS_DTYPE)
+            copy = False
+            # TODO: Why do we treat this differently from integer dtypes?
+
+        elif is_timedelta64_dtype(data):
+            warnings.warn("Passing timedelta64-dtype data to {cls} is "
+                          "deprecated, will raise a TypeError in a future "
+                          "version".format(cls=cls.__name__),
```

Comment:
Should this deprecation also be done in `to_datetime`?

Reply:
Yes.  Like with the sequence_to_td64ns and to_timedelta, the idea is to get a single non-circular implementation in place, then have both `DatetimeIndex.__new__` and `to_datetime` dispatch to it.